File: services/api/src/issue/issue.controller.ts
import { Controller, Get, Post, Put, Param, Body, Query } from '@nestjs/common';
import { IssueService } from './issue.service';
import { CurrentUser, AuthPrincipal } from '../common/decorators/current-user.decorator';
import { ProjectResourceAccessPolicy } from '../common/access/project-resource-access-policy';
import { CreateIssueDto, UpdateIssueDto, IssueTransitionDto } from './dto';

@Controller('projects/:projectId/issues')
export class IssueController {
  constructor(
    private readonly issueService: IssueService,
    private readonly accessPolicy: ProjectResourceAccessPolicy,
  ) {}

  @Post()
  async create(@Param('projectId') projectId: string, @Body() body: CreateIssueDto, @CurrentUser() user: AuthPrincipal) {
    await this.accessPolicy.requireProject(projectId, user.userId, user.workspaceId!, 'TESTER');
    return this.issueService.createIssue(projectId, body, user.userId);
  }

  @Post('from-failed-test')
  async createFromFailedTest(
    @Param('projectId') projectId: string,
    @Query('testRunCaseId') testRunCaseId: string,
    @Query('deploymentId') deploymentId: string,
    @Query('testCaseTitle') testCaseTitle: string,
    @CurrentUser() user: AuthPrincipal,
  ) {
    await this.accessPolicy.requireProject(projectId, user.userId, user.workspaceId!, 'TESTER');
    return this.issueService.createFromFailedTest(testRunCaseId, projectId, deploymentId, testCaseTitle || null, user.userId);
  }

  @Get()
  async list(
    @Param('projectId') projectId: string,
    @Query('status') status?: string,
    @Query('priority') priority?: string,
    @Query('assigneeId') assigneeId?: string,
    @CurrentUser() user?: AuthPrincipal,
  ) {
    await this.accessPolicy.requireProject(projectId, user!.userId, user!.workspaceId!, 'VIEWER');
    return this.issueService.listIssues(projectId, status, priority, assigneeId);
  }

  @Get(':id')
  async get(@Param('id') id: string, @CurrentUser() user: AuthPrincipal) {
    await this.accessPolicy.requireIssue(id, user.userId, user.workspaceId!, 'VIEWER');
    return this.issueService.getIssue(id);
  }

  @Put(':id')
  async update(@Param('id') id: string, @Body() body: UpdateIssueDto, @CurrentUser() user: AuthPrincipal) {
    await this.accessPolicy.requireIssue(id, user.userId, user.workspaceId!, 'DEVELOPER');
    return this.issueService.updateIssue(id, body);
  }

  @Put(':id/status')
  async transition(@Param('id') id: string, @Body() body: IssueTransitionDto, @CurrentUser() user: AuthPrincipal) {
    await this.accessPolicy.requireIssue(id, user.userId, user.workspaceId!, 'TESTER');
    return this.issueService.transition(id, body, user.userId);
  }
}

File: services/api-java-backup/src/main/java/com/launchly/issue/controllers/IssueController.java
package com.launchly.issue.controllers;

import com.launchly.common.security.AuthContext;
import com.launchly.issue.dto.*;
import com.launchly.issue.services.IssueService;
import jakarta.validation.Valid;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/projects/{projectId}/issues")
public class IssueController {
    private final IssueService issueService;

    public IssueController(IssueService issueService) {
        this.issueService = issueService;
    }

    @PostMapping
    @PreAuthorize("hasAnyRole('OWNER', 'ADMIN', 'DEVELOPER', 'TESTER')")
    public ResponseEntity<IssueResponse> create(@PathVariable String projectId,
                                                 @Valid @RequestBody CreateIssueRequest request) {
        return ResponseEntity.ok(issueService.createIssue(projectId, request, AuthContext.userId()));
    }

    @PostMapping("/from-failed-test")
    @PreAuthorize("hasAnyRole('OWNER', 'ADMIN', 'DEVELOPER', 'TESTER')")
    public ResponseEntity<IssueResponse> createFromFailedTest(@PathVariable String projectId,
                                                               @RequestParam String testRunCaseId,
                                                               @RequestParam String deploymentId,
                                                               @RequestParam(required = false) String testCaseTitle) {
        return ResponseEntity.ok(issueService.createFromFailedTest(
            testRunCaseId, projectId, deploymentId, testCaseTitle, AuthContext.userId()));
    }

    @GetMapping
    public ResponseEntity<List<IssueResponse>> list(@PathVariable String projectId,
                                                     @RequestParam(required = false) String status,
                                                     @RequestParam(required = false) String priority,
                                                     @RequestParam(required = false) String assigneeId) {
        return ResponseEntity.ok(issueService.listIssues(projectId, status, priority, assigneeId));
    }

    @GetMapping("/{id}")
    public ResponseEntity<IssueResponse> get(@PathVariable String projectId, @PathVariable String id) {
        return ResponseEntity.ok(issueService.getIssue(id));
    }

    @PutMapping("/{id}")
    @PreAuthorize("hasAnyRole('OWNER', 'ADMIN', 'DEVELOPER')")
    public ResponseEntity<IssueResponse> update(@PathVariable String projectId, @PathVariable String id,
                                                 @RequestBody UpdateIssueRequest request) {
        return ResponseEntity.ok(issueService.updateIssue(id, request));
    }

    @PutMapping("/{id}/status")
    @PreAuthorize("hasAnyRole('OWNER', 'ADMIN', 'DEVELOPER', 'TESTER')")
    public ResponseEntity<IssueResponse> transition(@PathVariable String projectId, @PathVariable String id,
                                                     @Valid @RequestBody TransitionRequest request) {
        return ResponseEntity.ok(issueService.transition(id, request, AuthContext.userId()));
    }
}
